src/common/configuration_store/mem_config_item.hpp

```diff
@@ -0,0 +1,108 @@
+#pragma once
+#include <stdint.h>
+#include <stdlib.h>
+#include <array>
+#include <tuple>
+#include <optional>
+#include "item_updater.hpp"
+using EepromKey = const char *;
+struct ConfigurationStoreStructure;
+template <class T = ConfigurationStoreStructure>
+class ConfigurationStore;
+
+template <class T>
+struct MemConfigBound {
+    std::optional<T> min;
+    std::optional<T> max;
+    T def;
+    bool validate(const T &data) const {
+        if (min.has_value() && max.has_value()) {
+            return (data >= min.value() && data <= max.value());
+        }
+        return true;
+    }
+};
+
+template <class T, class CovertTo = T>
+struct MemConfigItem {
+    EepromKey key;
+
+    T data;
+    T def_val;
+
+    void factory_reset() {
+        data = def_val;
+    }
+    operator T const &() const {
+        return data;
+    }
+
+    void Init(T new_data);
+
+    void Set(T new_data);
+    T Get() {
+        return data;
+    }
+
+    constexpr MemConfigItem(EepromKey key, const T def_val)
+        : key(key)
+        , data(def_val)
+        , def_val(def_val) {}
+};
+
+template <class T, size_t SIZE>
+struct MemConfigItem<std::array<T, SIZE>> {
+    EepromKey key;
+
+    std::array<T, SIZE> data;
+    T def_val;
+
+    void factory_reset() {
+        data.fill(def_val);
+    }
+
+    operator std::array<T, SIZE> const &() const {
+        return data;
+    }
+    void Init(std::array<T, SIZE> new_data);
+
+    void Set(std::array<T, SIZE> new_data);
+
+    const std::array<T, SIZE> &Get() {
+        return data;
+    }
+
+    constexpr MemConfigItem(EepromKey key, const T def_val)
+        : key(key)
+        , def_val(def_val) {
+        data.fill(def_val);
+    }
+};
+
+template <size_t SIZE>
+struct MemConfigItem<std::array<char, SIZE>> {
+    EepromKey key;
+    std::array<char, SIZE> data;
+    const char *def_val;
+
+    void factory_reset() {
+        strcpy(data.data(), def_val);
+    }
+
+    operator std::array<char, SIZE> const &() const {
+        return data;
+    }
+    void Init(std::array<char, SIZE> new_data);
+
+    void Set(std::array<char, SIZE> new_data);
+
+    const std::array<char, SIZE> &Get() {
+        return data;
+    }
+
+    constexpr MemConfigItem(EepromKey key, const char *def_val)
+        : key(key)
+        , def_val(def_val) {
+        factory_reset();
```

Comment:
/nitpick/
Can we please rename this to something like "set_to_default"?
Everytime I read "factory_reset" I though "OMG why does he do factory reset here?! 😅"